src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesCcm.Unix.cs

```diff
@@ -0,0 +1,211 @@
+            CheckParameters(nonce.Length, tag.Length);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+
+            int keyLength = _keyHandle.Key.Length * 8;
+            using (SafeEvpCipherCtxHandle ctx = Interop.Crypto.EvpCipherCreatePartial(
+                    GetCipher(keyLength),
+                    keyLength,
+                    0))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(ctx);
+
+                if (!Interop.Crypto.EvpAesCcmSetNonceLength(ctx, nonce.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesCcmSetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherSetKeyAndIV(ctx, ref MemoryMarshal.GetReference(_keyHandle.Key), ref MemoryMarshal.GetReference(nonce), 0 /*decrypting*/))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (associatedData.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref nullRef, ciphertext.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref MemoryMarshal.GetReference(associatedData), associatedData.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+                }
+
+                int plaintextBytesWritten;
+                if (ciphertext.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref MemoryMarshal.GetReference(plaintext), out plaintextBytesWritten, ref MemoryMarshal.GetReference(ciphertext), ciphertext.Length))
+                    {
+                        throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+                    }
+                }
+                else
+                {
+                    byte nonEmptyArray = 123;
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nonEmptyArray, out plaintextBytesWritten, ref nonEmptyArray, ciphertext.Length))
+                    {
+                        throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+                    }
+                }
+
+                if (plaintextBytesWritten != plaintext.Length)
+                {
+                    // this shouldn't happen
+                    throw new CryptographicException();
+                }
+
+                // note: no call to EvpCipherFinalEx - it will always fail
```

Comment:
I'd write this more as

// The OpenSSL documentation says not to call EvpCipherFinalEx for CCM, and calling it will report failure.

The comment as-is makes it sound sort of like "this doesn't work and I don't know why"

Reply:
This doesn't sound right. according to to openssl docs, EvpCipherFinalEx is where the tag gets checked. If you're seeing failures, that sounds like you have a bug.

Reply:
it is described in one of the example pages but not on the API itself. Will add a comment pointing to that

Reply:
https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption#Authenticated_Decryption_using_CCM_mode

> The tag verify is performed when you call the final EVP_DecryptUpdate and is reflected by the return value: there is no call to EVP_DecryptFinal.

Reinforced by the CCM handler:

```C
    /* EVP_*Final() doesn't return any data */
    if (in == NULL && out != NULL)
        return 0;
```